docs/exceptions.md

```diff
@@ -0,0 +1,266 @@
+these to be regular errors.
+
+### Identifying error class
+
+Can the execution of the error codepath be excluded by code review and static
+verification (more precise types?)
+
+* Yes: it is a programmer mistake
+* No: it is a regular error
+
+## Is the code 100% pure or potentially impure?
+
+### Impure code
+
+The code is considered impure when it's an `IO` action, a function that returns
+an `IO` action, or similar. The code can be _potentially_ impure when it is
+written in an abstract monad that can be instantiated to `IO`.
+
+Definitely impure:
+
+```
+readFile :: FilePath -> IO String
+```
+
+Potentially impure:
+
+```
+lookupThrow :: MonadThrow m => Map k v -> k -> m v
+```
+
+### Pure code
+
+Pure code is one that does not use `IO` operations. For example:
+
+```
+lookupMaybe :: Map k v -> k -> Maybe v
+```
+
+Notice that while it is conceivable that we have a `Map` of `IO` actions, and
+instantiate `lookupMaybe` to `Map k (IO ()) -> k -> Maybe (IO ())`, the
+instantiation does not inspect/use these `IO` actions, so it's not considered
+potentially impure.
+
+### Identifying code purity
+
+Is there an instantiation of type parameters that would mean that there are
+`IO` actions inspected/used in the code?
+
+* Yes: it is potentially impure code
+* No: it is 100% pure code
+
+## Error handling practices
+
+### Pure code, programmer mistakes
+
+Do *not*:
+
+* do this often, as if it's okay
+* use `undefined`
+* use non-exhaustive pattern matching
+
+Do:
+
+* try to use types to avoid the need in the first place
+* comment extensively (invariants and precondition, reasoning)
+* use the `error` function (or `impureThrow` with custom exception)
+* use the `HasCallStack` feature
+
+Before writing code that allows programmer mistakes, consult with colleagues how
+it would be possible to prevent them statically. There might be tricks that
+you're not aware of. Always make this trade-off conciously and responsibly.
+
+(On the other hand, if static guarantees require GADTs or type families, perhaps
+it's better to not overcomplicate code. Use your judgement.)
+
+DISCUSSION: Should we create a synonym `bug = impureThrow` in Universum? This
+would make the intention more clear.
+
+### Pure code, regular errors
+
+Do *not*:
+
+* use `error` or `impureThrow`
+* use `MonadFail`
+* return `Either Text`
+
+Do:
+
+* return `Either ErrorADT`, `Maybe`
+* wrap the underlying (pure!) monad in `ExceptT` or `CatchT`
+* use `MonadError` or `MonadThrow`
+
+Consider parsing: it is pure, but we cannot make assumptions about the input. In
+this case we might want to use `ExceptT ParseError`. Or consider a lookup in a
+`Map`, where we don't know whether the key is present -- in this case we'd like
+to return `Maybe v`. In 100% pure code, use one of these ways to handle errors:
+
+* `Maybe`, `Either e`
+* `MaybeT`, `ExceptT e`
+* `CatchT`
+
+Avoid using `Text` with the error message in place of `e` -- create a proper
```

Comment:
Again, I disagree. Consider that we're writing a message handler. What `impureFunc` returns is either data or error message to user.  Why would one use adt here? We never `catch` this error.
```
listener msg = 
  either respondError respondSuccess $ impureFunc msg
```
    
Rephrasing, sometimes `Either A B` is not a "way of exception handling", but is a handy way to return a result. One more example: let `checkData` be a impure function that has possible outcomes:
* Exit with success ~ "data is alright"
* Exit with failure ~ "data is not alright because T"
* Exit with normal error ~ "database became inaccessible (if we communicate w/ it over web)" or "we tried to write a log to file, but it doesn't exist".
* Exit with programmer error ~ "something went really wrong and it's not expected"

What you seem to miss is that sometimes it's easier to represent 1/2 difference with `Either Text a` instead of a custom ADT like `DataVerSuccess | DataVerErrorX | DataVerErrorY | ...`.

Reply:
Defining an ADT in Haskell is cheap. You never know what the caller of your function might want to do with the `Left` case, perhaps it has a way to handle specific cases. You assume that the only thing done to the error is displaying it, which is a strong assumption, and may change with time.

To be clear, this rule applies to top-level functions, where you cannot know what the caller intends to do with the return value. Maybe we want to log the error as JSON rather than text, maybe we want to return it from a HTTP request (and assign an appropriate error code), maybe we want to display it in a different language (Japanese), etc...

Just be nice to the consumers of your API and don't return `Text`. Instead, return an `ADT` and a function (or a `Buildable` instance) to display the error.

Reply:
@int-index Ok, I support your point in fact, but I guess it should be explicitly stated in the document. It's not that straightforward and clear why do we ban using `Either` for returning non-exceptional failures.

Reply:
I tried to keep the guidelines to the point, without justifications, so people can use them as a quick reference to make a decision. Good thing that this GH discussion will remain and can be linked to in case there are questions.